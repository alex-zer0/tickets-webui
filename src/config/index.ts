const config = {
  apiUrl: 'http://localhost:8000/api',
  pusherTicketsChannel: 'tickets_channel',
  pusherTicketsEvent: 'tickets_events',
  pusherAppKey: '9a46d976311a4b4ae129',
  pusherCluster: 'eu',
};

export default config;
