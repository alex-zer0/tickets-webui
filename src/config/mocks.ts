export const fakeEvent = () => ({
  name: 'Event',
  url: '/event',
  bannerUrl: 'http://event-banner-url',
  date: '2020-10-10 00:00:00',
  location: 'Russia, Saint-Petersburg',
  description: 'Test Description',
  tickets: [],
});

export const fakeTickets = () => [
  {id: 1, name: 'VIP', available: 10},
  {id: 2, name: 'Standard', available: null},
  {id: 3, name: 'Alumnus', available: null},
];
