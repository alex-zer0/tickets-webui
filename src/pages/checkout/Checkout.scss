.Checkout {
    box-shadow: 0 1px 4px rgba(0,0,0,.3);
    width: 900px;
    margin: 0 auto;
}
