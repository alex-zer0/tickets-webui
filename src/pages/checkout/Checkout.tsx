import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Redirect } from 'react-router-dom';
import { Dispatch } from 'redux';
import Header from 'src/components/header/Header';
import { releaseEventTickets, confirmEventTickets, resetConfirmReferenceId } from 'src/store/actions/events';
import { EventInfo, EventsState, UserInfo, TicketFormEventTicket } from 'src/store/reducers/events';
import ConfirmModal from './modals/ConfirmModal';
import PersonalInfoForm from './personal-info-form/PersonalInfoForm';
import './Checkout.scss';

export interface CheckoutProps {
  event: EventInfo | null
  tickets: TicketFormEventTicket[] | null
  confirmReferenceId: number | null
  confirmEventTickets: (url: string, tickets: TicketFormEventTicket[], user: UserInfo) => void 
  releaseEventTickets: (url: string, tickets: TicketFormEventTicket[]) => void
  resetConfirmReferenceId: () => void
}
interface CheckoutState {
  showConfirmModal: boolean
  goToEvent: boolean
  isReleased: boolean
  referenceId: number | null
}

export class Checkout extends Component<CheckoutProps, CheckoutState> {
  public state: CheckoutState = {
    showConfirmModal: false,
    referenceId: null,
    goToEvent: false,
    isReleased: false,
  };

  public handleConfirm = async (user: UserInfo) => {
    const { event, tickets, confirmEventTickets } = this.props;
    if (!event || !tickets) {
      return;
    }
    confirmEventTickets(event.url, tickets, user);
  }

  public goBack = async () => {
    const { event, tickets, confirmReferenceId, releaseEventTickets, resetConfirmReferenceId } = this.props;
    if (!event) {
      return;
    }
    if (confirmReferenceId) {
      resetConfirmReferenceId();
    } else if (tickets) {
      releaseEventTickets(event.url, tickets);
    }
  }

  public componentWillUnmount() {
    this.goBack();
  }

  public render() {
    const { event, tickets, confirmReferenceId } = this.props;
    if (!event) {
      return null;
    }
    if (!tickets && !confirmReferenceId) {
      return <Redirect to={event.url} />;
    }

    const { name, bannerUrl } = event;
    return (
      <div className='Checkout'>
        <Header title={name} imageUrl={bannerUrl} pageTitle='Checkout' />
        {tickets && <PersonalInfoForm event={event} tickets={tickets} confirm={this.handleConfirm} goBack={this.goBack} />}
        {confirmReferenceId && <ConfirmModal event={event} referenceId={confirmReferenceId} />}
      </div>
    );
  }
}

function mapStateToProps(state: { eventsState: EventsState }) {
  return {
    event: state.eventsState.activeEvent,
    tickets: state.eventsState.checkoutTickets,
    confirmReferenceId: state.eventsState.confirmReferenceId
  };
}

function mapDispatchToProps(dispatch: Dispatch) {
  return {
    confirmEventTickets: (url: string, tickets: TicketFormEventTicket[], user: UserInfo) =>
      dispatch(confirmEventTickets(url, tickets, user)),
    releaseEventTickets: (url: string, tickets: TicketFormEventTicket[]) =>
      dispatch(releaseEventTickets(url, tickets)),
    resetConfirmReferenceId: () => dispatch(resetConfirmReferenceId()),
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(Checkout);
