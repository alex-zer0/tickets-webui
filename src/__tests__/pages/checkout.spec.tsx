import React from 'react';
import renderer from 'react-test-renderer';
import { fakeEvent } from '../../config/mocks';
import { Checkout, CheckoutProps } from '../../pages/checkout/Checkout';

describe('PAGES: Checkout', () => {
  it('renders w/o exceptions', () => {
    const props: CheckoutProps = {
      event: fakeEvent(),
      tickets: [],
      confirmReferenceId: null,
      confirmEventTickets: jest.fn(),
      releaseEventTickets: jest.fn(),
      resetConfirmReferenceId: jest.fn()
    };

    const tree = renderer
      .create(<Checkout {...props}/>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
