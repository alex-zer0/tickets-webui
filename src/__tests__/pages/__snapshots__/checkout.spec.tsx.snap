// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PAGES: Checkout renders w/o exceptions 1`] = `
<div
  className="Checkout"
>
  <header
    className="Header"
  >
    <img
      alt="nodejs"
      className="Header__image"
      src="http://event-banner-url"
    />
    <h1
      className="Header__title"
    >
      Event
       &gt; Checkout
    </h1>
  </header>
  <div
    className="PersonalInfoForm"
  >
    <h3>
      Confirm personal information
    </h3>
    <form
      onSubmit={[Function]}
    >
      <section>
        <div
          className="row"
        >
          <div
            className="form-group col-6"
          >
            <label>
              <span>
                First Name 
                <i
                  className="text-danger"
                >
                  *
                </i>
              </span>
              <input
                autoComplete="firstName"
                className="form-control"
                name="firstName"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                type="text"
                value=""
              />
            </label>
          </div>
          <div
            className="form-group col-6"
          >
            <label>
              <span>
                Last Name 
                <i
                  className="text-danger"
                >
                  *
                </i>
              </span>
              <input
                autoComplete="lastName"
                className="form-control"
                name="lastName"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                type="text"
                value=""
              />
            </label>
          </div>
        </div>
        <div
          className="row"
        >
          <div
            className="form-group col-6"
          >
            <label>
              <span>
                Email 
                <i
                  className="text-danger"
                >
                  *
                </i>
              </span>
              <input
                autoComplete="email"
                className="form-control"
                name="email"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                type="email"
                value=""
              />
            </label>
          </div>
          <div
            className="form-group col-6"
          >
            <label>
              <span>
                Phone Number 
                <i
                  className="text-danger"
                >
                  *
                </i>
              </span>
              <input
                autoComplete="tel"
                className="form-control"
                name="phone"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                type="text"
                value=""
              />
            </label>
          </div>
        </div>
        <div>
          <p>
            Event
          </p>
          <p>
            Invalid DateTime
          </p>
          <h5>
            Your tickets
          </h5>
        </div>
        <footer>
          <button
            className="btn btn-default"
            onClick={[Function]}
            type="button"
          >
            Back
          </button>
          <button
            className="btn btn-primary"
            disabled={true}
            type="submit"
          >
            Confirm
          </button>
        </footer>
      </section>
    </form>
  </div>
</div>
`;
